### DISCUSSION TITLE
Adding elements directly with the clipboard

### DISCUSSION THREAD

**Discussion by oriolgalceran** [qelectrotech/qelectrotech-source-mirror#306] [General]
Hi, I've read the wiki but I can't seem to understand if there is the possibility to paste element definitions directly from the clipboard, for example pasting something like this 

`  <definition type="element" width="45" height="65" hotspot_x="25" hotspot_y="45" orientation="dnnn">
    <names>
      <name lang="fr">Moteur asynchrone triphasé</name>
      <name lang="en">Three-phase induction motor</name>
    </names>
    <informations>Author: Xavier Guerrin
License: CC-BY
</informations>
    <description>
      
      <polygon x1="-20" y1="-40" x2="-20" y2="-20" x3="-14" y3="-14" antialias="false" closed="false"/>
      
      <line x1="0" y1="-40" x2="0" y2="-20" antialias="false"/>
      
      <polygon x1="20" y1="-40" x2="20" y2="-20" x3="14" y3="-14" antialias="false" closed="false"/>
      <circle x="-20" y="-20" diameter="40" antialias="true" style="filling:white;"/>
      
      <text x="-13.5" y="12" size="24" text="M"/>
      
      <terminal orientation="n" x="-20" y="-40"/>
      <terminal orientation="n" x="0"   y="-40"/>
      <terminal orientation="n" x="20"  y="-40"/>
    </description>
  </definition>`

directly from the clipboard to the schematic. Or do we have to add the element manually to our collection and then reference it? 

Thanks!!

**Comment by ispyisail** [comment]
Sorry this went unanswered for so long. Short version: **your second guess is the correct one — the element has to exist in a collection (or embedded in the project) first, and the diagram then references it.** You cannot paste a raw `<definition>` block onto a schematic.

## Why

Pasting does read XML from the clipboard — `DiagramView::paste()` takes the clipboard text and parses it — but the parsed document is handed to `Diagram::fromXml()`, which starts with:

```cpp
if (root.tagName() != QLatin1String("diagram")) {
    return false;
}
```

Your snippet's root tag is `<definition>`, so it is rejected immediately and nothing happens — which is presumably what you saw.

What QET itself puts on the clipboard when you copy from a schematic is a `<diagram>` fragment, and the elements inside it look roughly like:

```xml
<diagram …>
  <elements>
    <element type="embed://…" x="…" y="…" …/>
  </elements>
</diagram>
```

The `type` attribute is a **reference to where the element lives**, not the element itself. On paste, `ElementFactory::createElement()` resolves that reference and loads the definition from the collection (or from the project's embedded collection for `embed://`). The geometry — your `<polygon>`, `<circle>`, `<terminal>` entries — is never carried inline in a diagram. So there is nowhere in the paste path for a bare definition to go.

## What to do instead

A `<definition>` block is exactly the contents of a `.elmt` file, so save it as one and drop it in your user collection:

| OS | User collection folder |
|---|---|
| Linux | `~/.local/share/qelectrotech/elements/` |
| Windows | `C:\Users\<you>\AppData\Roaming\qelectrotech\elements\` |
| macOS | `~/Library/Application Support/qelectrotech/elements/` |

Then, **without restarting**: right-click in the Collections panel → **"Reload collections"**. Your element appears in the tree and you can drag it onto a schematic, or copy/paste it like any other.

(If you have set a custom collection path in the preferences, use that folder instead — it takes priority over the default above.)

## If you are doing this in bulk or from a script

I suspect the clipboard idea was a shortcut toward automation. If so, writing `.elmt` files directly into that folder **is** the supported route — QET reads them from disk, so generating a few hundred of them from a script and hitting "Reload collections" works fine. There is no need to go through the GUI for each one.

Two things worth knowing if you go that way:

- **`embed://`** — elements can also live inside a `.qet` project rather than in a shared collection. That is how a copied element travels between projects: the definition is embedded in the project file. If you want your parts to move with the project rather than depend on a local folder, that is the mechanism.
- The element editor can open an existing `.elmt` from disk, so you can build one by hand, save it, and check that it loads and renders before generating the rest.

Hope that is still useful two years on — and if you already worked it out, the answer is at least here for the next person who searches for it.

**Reply by oriolgalceran** [reply]
I guess AI finally got round to the stale issues lol - and that's rare

**Comment by ispyisail** [comment]
Better late than never
